File: sdks/typescript/src/services/incidents.ts
/**
 * Incidents API Service
 *
 * Handles incident-related API operations.
 */

import { AxiosInstance } from 'axios';
import { handleAPIError } from '../client';
import {
  Incident,
  IncidentList,
  IncidentSeverity,
  IncidentStatus,
  IncidentPhase,
  IncidentNote,
  IncidentTimelineEntry,
} from '../models';

/** Options for listing incidents */
export interface IncidentListOptions {
  query?: string;
  severity?: IncidentSeverity;
  status?: IncidentStatus;
  phase?: IncidentPhase;
  assignee?: string;
  timeRange?: string;
  sortBy?: string;
  sortOrder?: 'asc' | 'desc';
  limit?: number;
  offset?: number;
}

/** Options for creating an incident */
export interface IncidentCreateOptions {
  title: string;
  description: string;
  severity: IncidentSeverity;
  hosts?: string[];
  detections?: string[];
  tags?: string[];
  assignee?: string;
}

/** Options for updating an incident */
export interface IncidentUpdateOptions {
  title?: string;
  description?: string;
  severity?: IncidentSeverity;
  tags?: string[];
}

/** Options for updating incident status */
export interface IncidentUpdateStatusOptions {
  status: IncidentStatus;
  notes?: string;
}

/** Options for updating incident phase */
export interface IncidentUpdatePhaseOptions {
  phase: IncidentPhase;
  notes?: string;
}

/** Options for closing an incident */
export interface IncidentCloseOptions {
  resolution: string;
  lessonsLearned?: string;
}

/** Incidents API service */
export class IncidentsService {
  private readonly client: AxiosInstance;
  private readonly basePath = '/incidents';

  constructor(client: AxiosInstance) {
    this.client = client;
  }

  /**
   * List incidents with optional filtering
   */
  async list(options?: IncidentListOptions): Promise<IncidentList> {
    try {
      const params: Record<string, string | number> = {};

      if (options?.query) params.query = options.query;
      if (options?.severity) params.severity = options.severity;
      if (options?.status) params.status = options.status;
      if (options?.phase) params.phase = options.phase;
      if (options?.assignee) params.assignee = options.assignee;
      if (options?.timeRange) params.time_range = options.timeRange;
      if (options?.sortBy) params.sort_by = options.sortBy;
      if (options?.sortOrder) params.sort_order = options.sortOrder;
      if (options?.limit) params.limit = options.limit;
      if (options?.offset) params.offset = options.offset;

      if (!params.limit) params.limit = 50;

      const response = await this.client.get<IncidentList>(this.basePath, { params });
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Get a specific incident by ID
   */
  async get(incidentId: string): Promise<Incident> {
    try {
      const response = await this.client.get<Incident>(
        `${this.basePath}/${incidentId}`
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Create a new incident
   */
  async create(options: IncidentCreateOptions): Promise<Incident> {
    try {
      const response = await this.client.post<Incident>(this.basePath, options);
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Update an incident
   */
  async update(incidentId: string, options: IncidentUpdateOptions): Promise<Incident> {
    try {
      const response = await this.client.patch<Incident>(
        `${this.basePath}/${incidentId}`,
        options
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Update the status of an incident
   */
  async updateStatus(
    incidentId: string,
    options: IncidentUpdateStatusOptions
  ): Promise<Incident> {
    try {
      const response = await this.client.patch<Incident>(
        `${this.basePath}/${incidentId}/status`,
        options
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Update the phase of an incident
   */
  async updatePhase(
    incidentId: string,
    options: IncidentUpdatePhaseOptions
  ): Promise<Incident> {
    try {
      const response = await this.client.patch<Incident>(
        `${this.basePath}/${incidentId}/phase`,
        options
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Assign an incident to a user
   */
  async assign(incidentId: string, assignee: string): Promise<Incident> {
    try {
      const response = await this.client.patch<Incident>(
        `${this.basePath}/${incidentId}/assign`,
        { assignee }
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Add a note to an incident
   */
  async addNote(
    incidentId: string,
    note: string,
    visibility?: string
  ): Promise<IncidentNote> {
    try {
      const response = await this.client.post<IncidentNote>(
        `${this.basePath}/${incidentId}/notes`,
        { content: note, visibility: visibility ?? 'team' }
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * List notes for an incident
   */
  async listNotes(incidentId: string, limit?: number): Promise<IncidentNote[]> {
    try {
      const params: Record<string, number> = {};
      if (limit) params.limit = limit;

      const response = await this.client.get<{ notes: IncidentNote[] }>(
        `${this.basePath}/${incidentId}/notes`,
        { params }
      );
      return response.data.notes;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Get the timeline for an incident
   */
  async getTimeline(incidentId: string): Promise<IncidentTimelineEntry[]> {
    try {
      const response = await this.client.get<{ entries: IncidentTimelineEntry[] }>(
        `${this.basePath}/${incidentId}/timeline`
      );
      return response.data.entries;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Add hosts to an incident
   */
  async addHosts(incidentId: string, hostIds: string[]): Promise<Incident> {
    try {
      const response = await this.client.post<Incident>(
        `${this.basePath}/${incidentId}/hosts`,
        { host_ids: hostIds }
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Remove hosts from an incident
   */
  async removeHosts(incidentId: string, hostIds: string[]): Promise<Incident> {
    try {
      const response = await this.client.delete<Incident>(
        `${this.basePath}/${incidentId}/hosts`,
        { data: { host_ids: hostIds } }
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Add detections to an incident
   */
  async addDetections(incidentId: string, detectionIds: string[]): Promise<Incident> {
    try {
      const response = await this.client.post<Incident>(
        `${this.basePath}/${incidentId}/detections`,
        { detection_ids: detectionIds }
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Remove detections from an incident
   */
  async removeDetections(incidentId: string, detectionIds: string[]): Promise<Incident> {
    try {
      const response = await this.client.delete<Incident>(
        `${this.basePath}/${incidentId}/detections`,
        { data: { detection_ids: detectionIds } }
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Close an incident
   */
  async close(incidentId: string, options: IncidentCloseOptions): Promise<Incident> {
    try {
      const response = await this.client.post<Incident>(
        `${this.basePath}/${incidentId}/close`,
        options
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }

  /**
   * Search for incidents matching a query
   */
  async search(query: string, limit?: number): Promise<IncidentList> {
    try {
      const params: Record<string, string | number> = { query };
      if (limit) params.limit = limit;
      else params.limit = 50;

      const response = await this.client.get<IncidentList>(
        `${this.basePath}/search`,
        { params }
      );
      return response.data;
    } catch (error) {
      handleAPIError(error);
    }
  }
}

File: sdks/go/vsentinel/incidents.go
package vsentinel

import (
	"context"
)

// IncidentsService handles incident-related API calls.
type IncidentsService struct {
	client *Client
}

// IncidentListOptions represents options for listing incidents.
type IncidentListOptions struct {
	Query     string
	Severity  IncidentSeverity
	Status    IncidentStatus
	Phase     IncidentPhase
	Assignee  string
	TimeRange string
	SortBy    string
	SortOrder string
	Limit     int
	Offset    int
}

// toParams converts options to query parameters.
func (o *IncidentListOptions) toParams() map[string]string {
	params := make(map[string]string)

	if o.Query != "" {
		params["query"] = o.Query
	}
	if o.Severity != "" {
		params["severity"] = string(o.Severity)
	}
	if o.Status != "" {
		params["status"] = string(o.Status)
	}
	if o.Phase != "" {
		params["phase"] = string(o.Phase)
	}
	if o.Assignee != "" {
		params["assignee"] = o.Assignee
	}
	if o.TimeRange != "" {
		params["time_range"] = o.TimeRange
	}
	if o.SortBy != "" {
		params["sort_by"] = o.SortBy
	}
	if o.SortOrder != "" {
		params["sort_order"] = o.SortOrder
	}
	if o.Limit > 0 {
		params["limit"] = intToString(o.Limit)
	}
	if o.Offset > 0 {
		params["offset"] = intToString(o.Offset)
	}

	return params
}

// List retrieves a list of incidents.
func (s *IncidentsService) List(ctx context.Context, opts *IncidentListOptions) (*IncidentList, error) {
	var result IncidentList

	params := make(map[string]string)
	if opts != nil {
		params = opts.toParams()
	}

	if _, ok := params["limit"]; !ok {
		params["limit"] = "50"
	}

	err := s.client.get(ctx, "/incidents", params, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// Get retrieves a specific incident by ID.
func (s *IncidentsService) Get(ctx context.Context, incidentID string) (*Incident, error) {
	var result Incident

	err := s.client.get(ctx, "/incidents/"+incidentID, nil, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// CreateOptions represents options for creating an incident.
type IncidentCreateOptions struct {
	Title       string           `json:"title"`
	Description string           `json:"description"`
	Severity    IncidentSeverity `json:"severity"`
	Hosts       []string         `json:"hosts,omitempty"`
	Detections  []string         `json:"detections,omitempty"`
	Tags        []string         `json:"tags,omitempty"`
	Assignee    string           `json:"assignee,omitempty"`
}

// Create creates a new incident.
func (s *IncidentsService) Create(ctx context.Context, opts *IncidentCreateOptions) (*Incident, error) {
	var result Incident

	err := s.client.post(ctx, "/incidents", opts, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// UpdateOptions represents options for updating an incident.
type IncidentUpdateOptions struct {
	Title       string           `json:"title,omitempty"`
	Description string           `json:"description,omitempty"`
	Severity    IncidentSeverity `json:"severity,omitempty"`
	Tags        []string         `json:"tags,omitempty"`
}

// Update updates an incident.
func (s *IncidentsService) Update(ctx context.Context, incidentID string, opts *IncidentUpdateOptions) (*Incident, error) {
	var result Incident

	err := s.client.patch(ctx, "/incidents/"+incidentID, opts, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// UpdateStatusOptions represents options for updating an incident status.
type IncidentUpdateStatusOptions struct {
	Status IncidentStatus `json:"status"`
	Notes  string         `json:"notes,omitempty"`
}

// UpdateStatus updates the status of an incident.
func (s *IncidentsService) UpdateStatus(ctx context.Context, incidentID string, opts *IncidentUpdateStatusOptions) (*Incident, error) {
	var result Incident

	err := s.client.patch(ctx, "/incidents/"+incidentID+"/status", opts, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// UpdatePhaseOptions represents options for updating an incident phase.
type IncidentUpdatePhaseOptions struct {
	Phase IncidentPhase `json:"phase"`
	Notes string        `json:"notes,omitempty"`
}

// UpdatePhase updates the phase of an incident.
func (s *IncidentsService) UpdatePhase(ctx context.Context, incidentID string, opts *IncidentUpdatePhaseOptions) (*Incident, error) {
	var result Incident

	err := s.client.patch(ctx, "/incidents/"+incidentID+"/phase", opts, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// Assign assigns an incident to a user.
func (s *IncidentsService) Assign(ctx context.Context, incidentID string, assignee string) (*Incident, error) {
	var result Incident

	body := map[string]string{"assignee": assignee}

	err := s.client.patch(ctx, "/incidents/"+incidentID+"/assign", body, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// AddNote adds a note to an incident.
func (s *IncidentsService) AddNote(ctx context.Context, incidentID string, note string, visibility string) (*IncidentNote, error) {
	var result IncidentNote

	body := map[string]string{
		"content":    note,
		"visibility": visibility,
	}

	err := s.client.post(ctx, "/incidents/"+incidentID+"/notes", body, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// ListNotes lists notes for an incident.
func (s *IncidentsService) ListNotes(ctx context.Context, incidentID string, limit int) ([]IncidentNote, error) {
	var result struct {
		Notes []IncidentNote `json:"notes"`
	}

	params := map[string]string{}
	if limit > 0 {
		params["limit"] = intToString(limit)
	}

	err := s.client.get(ctx, "/incidents/"+incidentID+"/notes", params, &result)
	if err != nil {
		return nil, err
	}

	return result.Notes, nil
}

// GetTimeline retrieves the timeline for an incident.
func (s *IncidentsService) GetTimeline(ctx context.Context, incidentID string) ([]TimelineEntry, error) {
	var result struct {
		Entries []TimelineEntry `json:"entries"`
	}

	err := s.client.get(ctx, "/incidents/"+incidentID+"/timeline", nil, &result)
	if err != nil {
		return nil, err
	}

	return result.Entries, nil
}

// AddHosts adds hosts to an incident.
func (s *IncidentsService) AddHosts(ctx context.Context, incidentID string, hostIDs []string) (*Incident, error) {
	var result Incident

	body := map[string][]string{"host_ids": hostIDs}

	err := s.client.post(ctx, "/incidents/"+incidentID+"/hosts", body, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// RemoveHosts removes hosts from an incident.
func (s *IncidentsService) RemoveHosts(ctx context.Context, incidentID string, hostIDs []string) (*Incident, error) {
	var result Incident

	body := map[string][]string{"host_ids": hostIDs}

	err := s.client.delete(ctx, "/incidents/"+incidentID+"/hosts", body, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// AddDetections adds detections to an incident.
func (s *IncidentsService) AddDetections(ctx context.Context, incidentID string, detectionIDs []string) (*Incident, error) {
	var result Incident

	body := map[string][]string{"detection_ids": detectionIDs}

	err := s.client.post(ctx, "/incidents/"+incidentID+"/detections", body, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// RemoveDetections removes detections from an incident.
func (s *IncidentsService) RemoveDetections(ctx context.Context, incidentID string, detectionIDs []string) (*Incident, error) {
	var result Incident

	body := map[string][]string{"detection_ids": detectionIDs}

	err := s.client.delete(ctx, "/incidents/"+incidentID+"/detections", body, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// CloseOptions represents options for closing an incident.
type IncidentCloseOptions struct {
	Resolution      string `json:"resolution"`
	LessonsLearned  string `json:"lessons_learned,omitempty"`
}

// Close closes an incident.
func (s *IncidentsService) Close(ctx context.Context, incidentID string, opts *IncidentCloseOptions) (*Incident, error) {
	var result Incident

	err := s.client.post(ctx, "/incidents/"+incidentID+"/close", opts, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}

// Search searches for incidents matching a query.
func (s *IncidentsService) Search(ctx context.Context, query string, limit int) (*IncidentList, error) {
	var result IncidentList

	params := map[string]string{
		"query": query,
	}
	if limit > 0 {
		params["limit"] = intToString(limit)
	} else {
		params["limit"] = "50"
	}

	err := s.client.get(ctx, "/incidents/search", params, &result)
	if err != nil {
		return nil, err
	}

	return &result, nil
}